File: solutions/ocaml/bob/1/bob.ml
open Core.Std

let question phrase =
  let len = String.length phrase - 1 in
    match String.index phrase '?' with
      | None -> false
      | Some i -> i = len

let shout phrase =
  let up = String.uppercase phrase in
    up = phrase && up <> String.lowercase phrase

let silence phrase =
  String.strip phrase = ""

let response_for = function
  | phrase when silence phrase -> "Fine. Be that way!"
  | phrase when shout phrase -> "Whoa, chill out!"
  | phrase when question phrase -> "Sure."
  | _ -> "Whatever."

File: solutions/erlang/bob/1/bob.erl
-module(bob).
-export([response_for/1]).

response_for(Phrase) ->
  respond(string:strip(Phrase)).

respond("") ->
  "Fine. Be that way!";
respond(Phrase) ->
  case {shout(Phrase), question(Phrase)} of
    {shout, _} -> "Whoa, chill out!";
    {_, question} -> "Sure.";
    _Otherwise -> "Whatever."
  end.

shout(Phrase) ->
  case {string:to_upper(Phrase), re:replace(Phrase, "[A-Za-z]", "")} of
    {Phrase, Phrase} -> statement;
    {Phrase, _} -> shout;
    {_, _} -> statement
  end.

question(Phrase) ->
  case re:run(Phrase, "\\?$") of
    {match, _Position} -> question;
    nomatch -> statement
  end.